src/lib.rs

```diff
@@ -766,16 +766,106 @@ pub struct PreparedQueuePair<'res> {
     rnr_retry: u8,
 }
 
+#[cfg(feature = "serde")]
+extern crate serde;
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Serialize};
+
+// We go through this rigeramol because serde does now know how to derive unions.
+
+/// A (serde) serializable representation of a `QueuePairEndpoint`.
+#[derive(Copy, Clone, Debug)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub struct PortableQPEndpoint {
+    num: u32,
+    lid: u16,
+    gid_subnet_prefix: u64,
+    gid_interface_id: u64,
```

Comment:
This feels weird — `ibv_gid` is a union for a reason, and I'm not sure it makes sense to always interpret it as being a `global`. For example, what if the user is using the `raw` variant and encoding multiple `u16`s in it instead?

Reply:
I think it probably makes more sense to use `ibv_gid.raw`, so I switched to that for the fixup commit. Given that it's a union, and that as far as I'm aware, any set of 128 bits is a valid `[u8; 16]`, it shouldn't matter under what type we store the union as, it just might not make sense to read. The same invariants (which value, raw or global) that needed to be upheld before the data was serialized need to be upheld when the data is deserialized.

Reply:
So, the issue we run into here is that if a `QueuePairEndpoint` is serialized and sent over the wire to a host with different endianness, that other host will end up using the _wrong_ numbers when accessing the fields through the subnet prefix/interface id pair. I'm not exactly sure what the right or expected behavior here is. In some sense, this question stems back to "why do you want endpoints to be serializeable?", since they sort of fundamentally aren't transferrable in a meaningful way to other hosts.

Reply:
I'm not sure how to handle the endian problem for a union. I'm looking where to find out which discriminant is being used, but I don't see where it's encoded. 

My understanding (and I probably am wrong, if you don't understand what I'm trying to do) is that computers need to exchange endpoints over the network(non RDMA) before they can communicate over RDMA. As a `QueuePairEndpoint` is what we need to handshake on, it makes sense to just send that.

Reply:
Ah, no, you're right, I'm just paging this all back in — it's been a while :sweat_smile: 

I'm honestly not sure what the intended function of this `union` is here. I've been reading around and information is scarce. I did find this on [this page](https://www.rdmamojo.com/2012/08/02/ibv_query_gid/):

> Every GID has 128 bits and its value is combined from two parts:
>
> - Higher 64 bits: subnet prefix. Value that identifies a set of endports which are managed by a common SM
> - Lower 64 bits: GID prefix. EUI-64 value that is configured by the SM

Which _suggests_ that the full 128 bits are actually in a well-defined format, and that we should _always_ exchange `raw` without doing any endianness business. Which then suggests that the union is only there to make it more convenient to look at these values separately. But presumably to actually read them you _may_ have to do endianness conversion (and presumably you should only ever set the entire `raw` in one go). This in turn suggests to me that we may want to implement our own `ibv_gid` that is _just_ `raw` (and not a `union`) and then (maybe) define read-only accessor methods for the subnet prefix and GID prefix that do the right endianness conversion.

In the context of this change, I think we can stick with just operating on `raw`, and the move away from `union` is just a suggestion for how we can improve the interface (by removing this ambiguity) if you have the time to give it a try :)

Reply:
I think that https://github.com/jonhoo/rust-ibverbs/pull/7/commits/3cc8d7915c94bc229543a9752370981003a4a586 works. I replace `ibv_gid` with a rust type, and this solves our serialization problem. It makes the ffi slightly awkward, but not by much, and it clarifies how we interpret `ibv_gid`.

Reply:
Nice, thank you! How did you figure out that big endian is the right way to read out these values? Is there a spec we can reference?

Reply:
Nice, thank you! How did you figure out that big endian is the right way to read out these values? Is there a spec we can reference?

Reply:
No clear spec, but I'm going off of the header files in `rdma-core`. `ibv_gid` is defined as 
```c
union ibv_gid {
  uint8_t raw[16];
  struct {
    __be64 subnet_prefix;
    __be64 interface_id;
  } global;
};
```
`__be64` is typedefed to `__bitwise __u64`. This is a [sparse](https://en.wikipedia.org/wiki/Sparse) flag. My understanding is that it informs people that it should be read as a big endian 64 bit integer, converting to a normal number with `be64_to_cpu(const __be64);`.

Reply:
(for my own later reference: https://github.com/linux-rdma/rdma-core/blob/6dfd71e2549c12142a1c11cfeeb1e314c2e31280/libibverbs/verbs.h#L65-L71)

Ah, that's a great catch! Yes, that looks exactly right — thanks for digging into it.